Odata-docs/webapi-8/fundamentals/query-options.md

```diff
@@ -0,0 +1,562 @@
+
+```csharp
+[EnableQuery(AllowedQueryOptions = AllowedQueryOptions.Filter | AllowedQueryOptions.OrderBy)]
+public IQueryable<Customer> Get()
+{
+    return Customers.AsQueryable<Customer>();
+}
+```
+
+## Configured vs allowed query options
+When registering OData services, we add query options configurations as shown below:
+
+```csharp
+.AddOData(
+    options => options.Select().OrderBy().Expand().Count().SetMaxTop(null)
+)
+```
+
+These are global configurations. If a query option is not enabled here, it cannot be enabled in the controller.
+
+In the above example, we did not enable the `$filter` query option. If we configure a controller as shown below, the `$filter` will not be applied since we had not enabled `$filter` in the global configuration.
+
+```csharp
+[EnableQuery(AllowedQueryOptions = AllowedQueryOptions.Filter)]
+public IActionResult Get()
+{
+    return Ok(Customers);
+}
+```
+
+## Applying query options directly
+There are scenarios where we may be unable to use `EnableQuery` attribute. For example, if our controller fetches data from multiple data sources and not just a single `IQueryable`, we may want to control how the query options are applied and process the data before it's returned by the controller.
+
+We use the [ODataQueryOptions.ApplyTo](/dotnet/api/microsoft.aspnet.odata.query.odataqueryoptions.applyto) method to apply the required query options.
+
+[ODataQueryOptions<`TEntity`>](/dotnet/api/microsoft.aspnet.odata.query.odataqueryoptions-1) can be used an argument in a controller method.
+
+```csharp
+public IQueryable<Customer> Get(ODataQueryOptions<Customer> options)
+{
+    IQueryable results = options.ApplyTo(Customers.AsQueryable());
+
+    return results as IQueryable<Customer>;
+}
+```
+
+In the example above, the [ApplyTo](/dotnet/api/microsoft.aspnet.odata.query.odataqueryoptions.applyto) method applies to all query options.
+
+We call ApplyTo for each individual query as shown in the sub-sections below.
+
+### SelectExpand query
+We call `ApplyTo` on the `SelectExpand` property as follows:
+
+```csharp
+public IQueryable<Customer> Get(ODataQueryOptions<Customer> options)
+{
+    IQueryable results = options.SelectExpand.ApplyTo(Customers.AsQueryable(), new ODataQuerySettings());
+
+    return results as IQueryable<Customer>;
+}
+```
+
+`SelectExpand` handles both `$select` and `$expand` query options.
+
+The `options.SelectExpand` property is an instance of `SelectExpandQueryOption`. It will be set when we make the request below:
+
+```http
+GET http://localhost:6285/odata/Customers?$expand=Orders
+```
+
+If we don't want to pass the query option in the request, but we want the `Orders` property to be expanded, we can initialize the `SelectExpandQueryOption` and set it in the request.
+
+```csharp
+public IQueryable<Customer> Get(ODataQueryOptions<Customer> options)
+{
+    options.Request.ODataFeature().SelectExpandClause = new SelectExpandQueryOption(null, "Orders", options.Context,
```

Comment:
@KenitoInc Food for thought and obviously not a deal breaker... One thing I have found to be useful with code snippets is named parameters. Since the reader doesn't have the library reference when reading the doc, it helps answer some of the questions they might have about the shared code. Like in your case, it'll not be immediately obvious what `null` represents. But if named parameters were used, it'd be self-documenting:
```csharp
options.Request.ODataFeature().SelectExpandClause = new SelectExpandQueryOption(
    select: null,
    expand: "Orders",
    context: options.Context,
    queryOptionParser: new ODataQueryOptionParser(
        model: options.Context.Model,
        targetEdmType: options.Context.NavigationSource.EntityType(),
        targetNavigationSource: options.Context.NavigationSource,
        queryOptions: new Dictionary<string, string> { { "$expand", "Orders" } },
        container: options.Context.RequestContainer)).SelectExpandClause;
```